File: OTM/simulated_annealing.py
import math
import random
from utils import ConstructiveAlgorithm, RandomSolutions
from classes_solucao import KPFSolution

def get_random_neighbor(solution):
    """Gera um vizinho aleatório por 1-flip (inversão de um item)."""
    neighbor = solution.clone()
    index = random.randint(0, len(solution.x) - 1)
    neighbor.toggle_item(index)
    return neighbor

def simulated_annealing(problem, ASmax=100, T_min=1e-3, T_init=100.0, alpha=0.95):
    """
    Simulated Annealing para o problema KPFS.
    
    Args:
        problem: instância de KPFSProblem.
        ASmax: número máximo de iterações por temperatura.
        T_min: temperatura mínima (critério de parada).
        T_init: temperatura inicial.
        alpha: fator de resfriamento (ex: 0.95).
    
    Returns:
        Melhor solução encontrada (KPFSolution).
    """
    # Geração inicial com RCL
    initial_solutions = RandomSolutions(problem, size=10, randomness=0.5)
    s_star = ConstructiveAlgorithm(initial_solutions)
    s_current = s_star.clone()
    T = T_init

    while T > T_min:
        for _ in range(ASmax):
            s_neighbor = get_random_neighbor(s_current)
            delta = s_current.objective_value() - s_neighbor.objective_value()

            if delta <= 0:  # s_neighbor é melhor ou igual
                s_current = s_neighbor
                if s_current.objective_value() > s_star.objective_value():
                    s_star = s_current.clone()
            else:
                prob = math.exp(-delta / T)
                if random.random() < prob:
                    s_current = s_neighbor  # aceita pior com certa probabilidade

        T *= alpha  # resfriamento

    return s_star

File: OTM/simulated_annealing.cpp
#include "simulated_annealing.hpp"
#include <vector>
#include "utils.hpp"
#include <memory>
#include <random>
#include <algorithm>
#include <cstdlib>
#include <ctime>
#include <cmath>


KPFSolution Get_Random_Neighbour(const KPFSolution& solution, std::mt19937& gen) {

    KPFSolution neighbour = solution.clone();
    const auto& problem = solution.problem();
    size_t num_items = problem.items().size();

    // Escolha um índice aleatório válido
    std::uniform_int_distribution<> distrib(0, num_items - 1);
    int index = distrib(gen);
    // Inverte o estado de um item, conseguindo, assim, um vizinho
    neighbour.toggleItem(index);
    return neighbour;
}

KPFSolution Simulated_Annealing(KPFSProblem problem, int ASmax, float T_min, float T_init, float alpha, int relaxation){
    
    // Geração de um número aleatório
    random_device rd;
    mt19937 gen(rd());
    
    uniform_real_distribution<float> uni01(0.0f, 1.0f);

    // Uso de algoritmo construtivo para criar uma solução
    KPFSolution constructed = ConstructiveAlgorithm(problem, 1, gen);

    auto s_star_ptr    = make_unique<KPFSolution>(constructed.clone());
    auto s_current_ptr = make_unique<KPFSolution>(constructed.clone());

    float T = T_init;

    // Enquanto a temperatura for maior que a temperatura mínima 
    while (T > T_min) {
        int iterations_no_improvement = 0;
        for (int iter = 0; iter < ASmax; ++iter) {
            // Ao alcançar um número de iterações sem melhoria igual ao valor de relaxation
            // interrompemos a execução
            if (iterations_no_improvement >= relaxation) break;

            // Encontra um vizinho aleatório à solução atual
            KPFSolution neighbor = Get_Random_Neighbour(*s_current_ptr, gen);

            float obj_current = s_current_ptr->objectiveValue();
            float obj_neigh   = neighbor.objectiveValue();

            // Cálculo da diferença entre valores da função objetivo
            float delta = obj_current - obj_neigh;

            // Se a solução vizinho for melhor:
            if (delta <= 0.0f) {
                // Solução vizinha torna-se a atual
                s_current_ptr = make_unique<KPFSolution>(move(neighbor));
                // Se a solução atual for a melhor até o momento
                if (s_current_ptr->objectiveValue() > s_star_ptr->objectiveValue()) {
                    s_star_ptr = make_unique<KPFSolution>(*s_current_ptr);
                    iterations_no_improvement = 0;
                }
                // Caso contrário:
            } else {
                iterations_no_improvement += 1;
                float prob = exp(-delta / T);
                // Caso o valor aleatório for menor que essa probabilidade a solução vizinha torna-se a atual
                if (uni01(gen) < prob) {
                    s_current_ptr = make_unique<KPFSolution>(move(neighbor));
                }
            }
        }
        // Atualizamos a temperatura
        T *= alpha;
    }

    return *s_star_ptr;
}